File: min-stack/min-stack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minStack = []
        

    def push(self, val: int) -> None:
        
        if len(self.stack)==0:
            self.minStack.append(val)
        else:
            if self.minStack[-1]>val:
                self.minStack.append(val)
            else:
                self.minStack.append(self.minStack[-1])
        self.stack.append(val)
        return   

    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: min-stack/min-stack.java
class MinStack {
    
    private List<Integer> stack;
    private List<Integer> minStack;
    private int top;
    
    public MinStack() {
        this.stack = new ArrayList<>();
        this.minStack = new ArrayList<>();
        this.top = -1;
    }
    
    public void push(int val) {
        this.stack.add(val);
        if(this.minStack.size()>0){
            if(this.minStack.get(this.top) > val){
                this.minStack.add(val);
            }
            else{
                this.minStack.add(this.minStack.get(this.top));
            }
        }
        else{
            this.minStack.add(val);
        }
        this.top++;
    }
    
    public void pop() {
        this.stack.remove(this.top);
        this.minStack.remove(this.top);
        this.top--;
    }
    
    public int top() {
        return this.stack.get(this.top);
    }
    
    public int getMin() {
        return this.minStack.get(this.top);
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */